### PR TITLE
Label Search for Multiple Targets

### LINKED ISSUE #90: No way to balance labels per entity
If the labels are not very balanced and we want to limit the number of examples per entity, sometimes all of the same label are returned for an entity.

### LINKED ISSUE #119: Label Search for Multiple Targets
##### Label Maker
The `labeling_function` in a label maker can be a list of user-defined labeling functions.

```python
lm = cp.LabelMaker(
    ...
    labeling_function=[a, b],
)
```
##### Label Search
The search parameter for the number of examples should return labels from each labeling function. For example, in each entity, this search will return three examples from labeling functions `a` and `b`.

```python
lm.search(df, num_examples_per_instance=3)
```

```
customer_id         cutoff_time      a      b
          1 2014-01-01 00:45:30   True  False
          1 2014-01-01 01:45:30   True  False
          1 2014-01-01 02:45:30  False   True
```

##### Testing and QA Plan

The tests should result in 100% code coverage. Each unit test must use more than one labeling function. The search parameter for the number of examples must include the following test cases:
* The search parameter is a numeric value and returns multiple targets.
* The search parameter is a label dictionary. This is expected to raise an error.

### PR DESCRIPTION
**Pull request by jeff-hernandez**
This closes #119 by supporting more than one labeling function during the search process. This also resolves #90 by supporting search processes based on label values.

### CODE REVIEW

File: composeml/data_slice.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+class Context:
+    """Metadata for data slice."""
+
+    def __init__(self, gap=None, window=None, slice_number=None, target_entity=None, target_instance=None):
+        """Metadata for data slice.
+
+        Args:
+            gap (tuple) : Start and stop time for gap.
+            window (tuple) : Start and stop time for window.
+            slice (int) : Slice number.
```
[the remark this answers is not in the document]
Comment (jeff-hernandez) [reply composeml/data_slice.py]: Yes, updated docstring.

File: composeml/label_search.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+    def is_valid_labels(self, labels):
+        """Whether the label values are not null."""
+        return not any(map(isnull, labels.values()))
+
+    def reset_count(self):
+        """Reset the internal count of actual labels."""
+        self.actual_count = 0
+
+    def update_count(self, labels):
```
Comment (rwedge) [inline composeml/label_search.py:35]: does this method need the `labels` parameter?

File: composeml/label_maker.py [hunk 3]
```diff
@@ -254,98 +352,65 @@ def search(self,
...
-        bar_format = "Elapsed: {elapsed} | Remaining: {remaining} | "
-        bar_format += "Progress: {l_bar}{bar}| "
-        bar_format += self.target_entity + ": {n}/{total} "
-        total = len(df.groupby(self.target_entity))
-        finite_examples_per_instance = num_examples_per_instance > -1 and num_examples_per_instance != float('inf')
+        assert self.labeling_function, 'missing labeling function(s)'
 
-        if finite_examples_per_instance:
-            total *= num_examples_per_instance
+        if self.window_size is None and gap is None:
```
Comment (rwedge) [inline composeml/label_maker.py:371]: this check happens in a few places now. do we need multiple checks?

Comment (jeff-hernandez) [reply composeml/label_maker.py:371]: Yes, this check should only happen in `search` and `slice` methods, but I can refactor into a reusable function.

File: composeml/label_search.py (hunk 2 above)
Comment (jeff-hernandez) [reply composeml/label_search.py:35]: Search classes do require this parameter as an [interface](https://github.com/FeatureLabs/compose/pull/130/files#diff-22bf41f653d21f0ca3dd15dadf5e01e9R305) in the search process.

```python
search.update_count(labels)
```

The intention is to make an interface requirement, but not a requirement for inherited methods to use the parameter. The example search class does not need this parameter, whereas the label search class does.

**Comment by jeff-hernandez** [comment]
@rwedge the Counter class worked out well. Thanks for the suggestion.

File: composeml/label_maker.py [hunk 4]
```diff
@@ -254,98 +356,61 @@ def search(self,
         """Searches the data to calculates labels.
 
         Args:
-            df (DataFrame) : Data frame to search and extract labels.
-            num_examples_per_instance (int) : Number of examples per unique instance of target entity.
-            minimum_data (str) : Minimum data before starting search. Default value is first time of index.
-            gap (str or int) : Time between examples. Default value is window size.
+            df (DataFrame): Data frame to search and extract labels.
+            num_examples_per_instance (int): Number of examples per unique instance of target entity.
```
Comment (rwedge) [inline composeml/label_maker.py:360]: this should be updated to explain how to pass a dict

File: composeml/tests/test_label_maker.py [hunk 5]
```diff
@@ -261,7 +341,7 @@ def test_search_offset_mix_7(transactions, total_spent_fn):
     given_labels = lm.search(
         transactions,
         num_examples_per_instance=float('inf'),
-        verbose=False,
+        verbose=True,
```
Comment (rwedge) [inline composeml/tests/test_label_maker.py:344]: is this test meant to test `verbose`?

Comment (jeff-hernandez) [reply composeml/tests/test_label_maker.py:344]: No, the `versbose` parameter is not needed in the test. I can remove it.

**Comment by jeff-hernandez** [comment]
I think that can be tested when we update the label description in #122.

File: composeml/label_maker.py [hunk 6]
```diff
@@ -206,41 +207,141 @@ def slice(self, df, num_examples_per_instance, minimum_data=None, gap=None, drop
...
+                    self.target_entity: entity_id,
+                    'cutoff_time': ds.context.window[0],
+                    **labels,
+                })
+
+                search.update_count(labels)
+                if search.is_finite: progress_bar.update(n=1)
+                if search.is_complete: break
+
+            n = missing_examples(entity_count + 1) if search.is_finite else 1
```
Comment (rwedge) [inline composeml/label_maker.py:306]: can we add some short comments summarizing these progress bar steps?

Comment (jeff-hernandez) [reply composeml/label_maker.py:309]: Yes, I added some comments about the progress bar steps.

File: composeml/data_slice.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
+import pandas as pd
+
+
+class Context:
```
Comment (kmax12) [inline composeml/data_slice.py:4]: What if we just called this `DataSliceMetadata` or `DataSliceContext`?

Comment (jeff-hernandez) [reply composeml/data_slice.py:4]: That works. I like `DataSliceContext`.

**Pull request merged**